File: plugins/filter/auto-wah/auto-wah.js
// Auto-Wah — envelope-following band-pass that opens / closes based on
// the input level. Loud notes push the filter up, quiet notes let it
// fall — same "quack" you hear on funky guitar / synth bass with a
// MuTron III. The Biquad coefficient maths is the standard
// Robert Bristow-Johnson cookbook (public domain).

function AutoWahProcessor() {
  this.env = 0;
  // Biquad state (Direct Form II Transposed).
  this.z1 = 0;
  this.z2 = 0;
}
AutoWahProcessor.parameterDescriptors = [
  { name: 'sensitivity', label: 'Sense',    defaultValue: 0.6, minValue: 0,    maxValue: 1 },
  { name: 'minFreq',     label: 'Min Hz',   defaultValue: 250, minValue: 60,   maxValue: 2000 },
  { name: 'maxFreq',     label: 'Max Hz',   defaultValue: 2500,minValue: 400,  maxValue: 8000 },
  { name: 'q',           label: 'Q',        defaultValue: 4,   minValue: 0.5,  maxValue: 12 },
  { name: 'attack',      label: 'Attack ms',defaultValue: 5,   minValue: 0.5,  maxValue: 200 },
  { name: 'release',     label: 'Rel ms',   defaultValue: 80,  minValue: 5,    maxValue: 500 }
];
AutoWahProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0];
  var output = outputs[0][0];
  if (!input || !output) return true;
  var sense = parameters.sensitivity[0];
  var minF = parameters.minFreq[0];
  var maxF = Math.max(minF + 50, parameters.maxFreq[0]);
  var q = parameters.q[0];
  var attackCoef = Math.exp(-1 / (sampleRate * parameters.attack[0] * 0.001));
  var releaseCoef = Math.exp(-1 / (sampleRate * parameters.release[0] * 0.001));

  for (var i = 0; i < input.length; i++) {
    var x = input[i];
    // Envelope: full-wave rectify then attack/release follower.
    var rect = Math.abs(x);
    var coef = rect > this.env ? attackCoef : releaseCoef;
    this.env = rect + (this.env - rect) * coef;

    // Map env (0..1-ish, capped at ~0.7 typical voice peak) to cutoff
    // in log frequency so a small env change near min-freq sweeps
    // similarly to one near max-freq.
    var drive = Math.min(1, this.env * (1 + sense * 8));
    var logMin = Math.log(minF);
    var logMax = Math.log(maxF);
    var cutoff = Math.exp(logMin + (logMax - logMin) * drive);

    // Bandpass biquad coefficients (constant 0 dB peak gain).
    var w0 = 2 * Math.PI * cutoff / sampleRate;
    var sinW = Math.sin(w0);
    var cosW = Math.cos(w0);
    var alpha = sinW / (2 * q);
    var b0 = alpha;
    var b1 = 0;
    var b2 = -alpha;
    var a0 = 1 + alpha;
    var a1 = -2 * cosW;
    var a2 = 1 - alpha;
    var nb0 = b0 / a0, nb1 = b1 / a0, nb2 = b2 / a0;
    var na1 = a1 / a0, na2 = a2 / a0;

    var y = nb0 * x + this.z1;
    this.z1 = nb1 * x - na1 * y + this.z2;
    this.z2 = nb2 * x - na2 * y;
    output[i] = y;
  }
  return true;
};

registerProcessor('auto-wah', AutoWahProcessor);

File: plugins/filter/auto-wah/AutoWah.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Auto-Wah (native port) — envelope-following bandpass biquad. Loud
// notes push cutoff up, quiet notes let it fall — MuTron III "quack"
// on guitar / vocal. Biquad form: Direct Form II Transposed.
public final class AutoWah implements VocalMonitorNativePlugin {
    private float env = 0f;
    private float z1 = 0f, z2 = 0f;
    private int sampleRate = 44100;
    private float sensitivity = 0.6f, minFreq = 250f, maxFreq = 2500f,
                  q = 4f, attack = 5f, release = 80f;

    @Override
    public void init(int sr) { this.sampleRate = sr; env = 0f; z1 = 0f; z2 = 0f; }

    @Override
    public String[] parameterNames() {
        return new String[] { "sensitivity", "minFreq", "maxFreq", "q", "attack", "release" };
    }
    @Override public float parameterMin(String n) {
        switch (n) {
            case "sensitivity": return 0f;
            case "minFreq":     return 60f;
            case "maxFreq":     return 400f;
            case "q":           return 0.5f;
            case "attack":      return 0.5f;
            case "release":     return 5f;
            default:            return 0f;
        }
    }
    @Override public float parameterMax(String n) {
        switch (n) {
            case "sensitivity": return 1f;
            case "minFreq":     return 2000f;
            case "maxFreq":     return 8000f;
            case "q":           return 12f;
            case "attack":      return 200f;
            case "release":     return 500f;
            default:            return 1f;
        }
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "sensitivity": return 0.6f;
            case "minFreq":     return 250f;
            case "maxFreq":     return 2500f;
            case "q":           return 4f;
            case "attack":      return 5f;
            case "release":     return 80f;
            default:            return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "sensitivity": return "Sense";
            case "minFreq":     return "Min Hz";
            case "maxFreq":     return "Max Hz";
            case "q":           return "Q";
            case "attack":      return "Attack ms";
            case "release":     return "Rel ms";
            default:            return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "sensitivity": sensitivity = v; break;
            case "minFreq":     minFreq = v; break;
            case "maxFreq":     maxFreq = v; break;
            case "q":           q = v; break;
            case "attack":      attack = v; break;
            case "release":     release = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float maxF = Math.max(minFreq + 50f, maxFreq);
        final float attackCoef = (float) Math.exp(-1.0 / (sampleRate * attack * 0.001));
        final float releaseCoef = (float) Math.exp(-1.0 / (sampleRate * release * 0.001));
        final float senseMul = 1f + sensitivity * 8f;
        final float logMin = (float) Math.log(minFreq);
        final float logMax = (float) Math.log(maxF);
        final float qLocal = q;
        final float twoPiOverSr = (float) (2.0 * Math.PI / sampleRate);
        float e = env, zA = z1, zB = z2;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float x = input[i];
            float rect = x < 0 ? -x : x;
            float coef = rect > e ? attackCoef : releaseCoef;
            e = rect + (e - rect) * coef;

            float drive = e * senseMul;
            if (drive > 1f) drive = 1f;
            float cutoff = (float) Math.exp(logMin + (logMax - logMin) * drive);

            float w0 = twoPiOverSr * cutoff;
            float sinW = (float) Math.sin(w0);
            float cosW = (float) Math.cos(w0);
            float alpha = sinW / (2f * qLocal);
            float a0 = 1f + alpha;
            float nb0 = alpha / a0;
            float nb2 = -alpha / a0;
            float na1 = -2f * cosW / a0;
            float na2 = (1f - alpha) / a0;

            float y = nb0 * x + zA;
            zA = -na1 * y + zB;
            zB = nb2 * x - na2 * y;
            output[i] = y;
        }
        env = e; z1 = zA; z2 = zB;
    }
}
